src/System.Tuples/src/System/ValueTuple/ValueTuple.cs

```diff
@@ -0,0 +1,1119 @@
+                comparer.Equals(Item4, objTuple.Item4) && comparer.Equals(Item5, objTuple.Item5) && comparer.Equals(Item6, objTuple.Item6) &&
+                comparer.Equals(Item7, objTuple.Item7) && comparer.Equals(Rest, objTuple.Rest);
+        }
+
+        int IComparable.CompareTo(object other)
+        {
+            if (other == null) return 1;
+
+            if (!(other is ValueTuple<T1, T2, T3, T4, T5, T6, T7, TRest>))
+            {
+                throw new ArgumentException(SR.ArgumentException_ValueTupleIncorrectType, nameof(other));
+            }
+
+            var objTuple = (ValueTuple<T1, T2, T3, T4, T5, T6, T7, TRest>)other;
+
+            int c = s_comparer1.Compare(Item1, objTuple.Item1);
+            if (c != 0) return c;
+
+            c = s_comparer2.Compare(Item2, objTuple.Item2);
+            if (c != 0) return c;
+
+            c = s_comparer3.Compare(Item3, objTuple.Item3);
+            if (c != 0) return c;
+
+            c = s_comparer4.Compare(Item4, objTuple.Item4);
+            if (c != 0) return c;
+
+            c = s_comparer5.Compare(Item5, objTuple.Item5);
+            if (c != 0) return c;
+
+            c = s_comparer6.Compare(Item6, objTuple.Item6);
+            if (c != 0) return c;
+
+            c = s_comparer7.Compare(Item7, objTuple.Item7);
+            if (c != 0) return c;
+
+            return s_comparerRest.Compare(Rest, objTuple.Rest);
+        }
+
+        int IStructuralComparable.CompareTo(object other, IComparer comparer)
+        {
+            if (other == null) return 1;
+
+            if (!(other is ValueTuple<T1, T2, T3, T4, T5, T6, T7, TRest>))
+            {
+                throw new ArgumentException(SR.ArgumentException_ValueTupleIncorrectType, nameof(other));
+            }
+
+            var objTuple = (ValueTuple<T1, T2, T3, T4, T5, T6, T7, TRest>)other;
+
+            int c = comparer.Compare(Item1, objTuple.Item1);
+            if (c != 0) return c;
+
+            c = comparer.Compare(Item2, objTuple.Item2);
+            if (c != 0) return c;
+
+            c = comparer.Compare(Item3, objTuple.Item3);
+            if (c != 0) return c;
+
+            c = comparer.Compare(Item4, objTuple.Item4);
+            if (c != 0) return c;
+
+            c = comparer.Compare(Item5, objTuple.Item5);
+            if (c != 0) return c;
+
+            c = comparer.Compare(Item6, objTuple.Item6);
+            if (c != 0) return c;
+
+            c = comparer.Compare(Item7, objTuple.Item7);
+            if (c != 0) return c;
+
+            return comparer.Compare(Rest, objTuple.Rest);
+        }
+
+        public override int GetHashCode()
+        {
+            // We want to have a limited hash in this case.  We'll use the last 8 elements of the tuple
+            ITupleInternal rest = (ITupleInternal)Rest;
+            if (rest.Size >= 8) { return rest.GetHashCode(); }
```

Comment:
On second thought, I think we're fine. `var vt = default(ValueTuple<int, int, int, int, int, int, int, int>);` won't pass the constructor check.

Because all the types that implement `ITupleInternal` are structs, you won't be able to set the `Rest` to something else.